src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java

```diff
@@ -843,38 +839,36 @@ void blockingSend(DatagramPacket p) throws IOException {
             if (!isBlocking())
                 throw new IllegalBlockingModeException();
 
-            ByteBuffer src = null;
-            try {
-                InetSocketAddress target;
-                synchronized (p) {
-                    int len = p.getLength();
-                    src = Util.getTemporaryDirectBuffer(len);
-
+            synchronized (p) {
+                int len = p.getLength();
+                ByteBuffer src = Util.getTemporaryDirectBuffer(len);
+                try {
                     // copy bytes to temporary direct buffer
                     src.put(p.getData(), p.getOffset(), len);
                     src.flip();
 
                     // target address
+                    InetSocketAddress target;
                     if (p.getAddress() == null) {
                         InetSocketAddress remote = remoteAddress();
                         if (remote == null) {
                             throw new IllegalArgumentException("Address not set");
                         }
-                        // set address/port to be compatible with long standing behavior
+                        // set address/port to be compatible with long-standing behavior
                         p.setAddress(remote.getAddress());
                         p.setPort(remote.getPort());
                         target = remote;
                     } else {
                         target = (InetSocketAddress) p.getSocketAddress();
                     }
-                }
-
-                // send the datagram (does not block)
-                send(src, target);
 
-            } finally {
-                if (src != null) Util.offerFirstTemporaryDirectBuffer(src);
+                    // send the datagram (does not block)
```

Comment:
Now this send is in the synchronized block; should this "does not block" be moved to the top level synchronized block?

Reply:
No, it's in the correct place.